examples/vision/edsr.py

```diff
@@ -0,0 +1,276 @@
+    plt.title(hr[i].shape)
+    plt.axis("off")
+
+"""
+# Low Resolution Images
+"""
+
+plt.figure(figsize=(10, 10))
+for i in range(9):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(lr[i].numpy().astype("uint8"))
+    plt.title(lr[i].shape)
+    plt.axis("off")
+
+"""
+# Normalization and usefull functions
+As, the paper mentioned we will subtract image by mean rgb of div2k dataset
+"""
+
+DIV2K_RGB_MEAN = np.array([0.4488, 0.4371, 0.4040]) * 255
+
+
+def normalize(x, rgb_mean=DIV2K_RGB_MEAN):
+    return (x - rgb_mean) / 127.5
+
+
+def denormalize(x, rgb_mean=DIV2K_RGB_MEAN):
+    return (x * 127.5) + rgb_mean
+
+
+def shuffle_pixels(scale):
+    return lambda x: tf.nn.depth_to_space(x, scale)
+
+
+# computes the peak signal-to-noise ratio, measures quality of image
+def PSNR(sr, hr):
+    psnr_value = tf.image.psnr(hr, sr, max_val=255)[0]
+    return psnr_value
+
+
+"""
+# Model Building(Baseline Model)
+1. 16 residual blocks
+2. filters=64
+3. No batchNormalization
+"""
+
+"""
+<img src = "https://miro.medium.com/max/1400/1*a1tQ4grku_ugTTe39IJzjA.png" width="500">
+"""
+
+
+def ResBlock(inputs):
+    x = tf.keras.layers.Conv2D(64, 3, padding="same", activation="relu")(inputs)
+    x = tf.keras.layers.Conv2D(64, 3, padding="same")(x)
+    x = tf.keras.layers.Add()([inputs, x])
+    return x
+
+
+def Upsampling(inputs, factor=2, **kwargs):
+    x = tf.keras.layers.Conv2D(64 * (factor**2), 3, padding="same", **kwargs)(inputs)
+    x = tf.keras.layers.Lambda(shuffle_pixels(scale=factor))(x)
+    x = tf.keras.layers.Conv2D(64 * (factor**2), 3, padding="same", **kwargs)(x)
+    x = tf.keras.layers.Lambda(shuffle_pixels(scale=factor))(x)
+    return x
+
+
+def EDSR():
+    input_layer = tf.keras.layers.Input(shape=(None, None, 3))
+    x = tf.keras.layers.Lambda(normalize)(input_layer)
+    x = x_new = tf.keras.layers.Conv2D(64, 3, padding="same")(x)
+
+    # 16 residual blocks
+    for _ in range(16):
+        x_new = ResBlock(x_new)
+
+    x_new = tf.keras.layers.Conv2D(64, 3, padding="same")(x_new)
+    x = tf.keras.layers.Add()([x, x_new])
+
+    x = Upsampling(x)
+    x = tf.keras.layers.Conv2D(3, 3, padding="same")(x)
+    output_layer = tf.keras.layers.Lambda(denormalize)(x)
+    return tf.keras.models.Model(input_layer, output_layer)
+
+
+edsr = EDSR()
+edsr.summary()
+
+"""
+# Training
+EDSR can take a long time to train, in the code below, but we will train for 5 epochs and
+20 steps for checking if the code is working alright. In practice, a larger steps value
+(over 200000) is required to get decent results.
+
+
+paper Suggest that learning rate of adam optimizer will be halfed after 2 × 10^5 steps.
+But as we are training for only 5 epochs we will change it to 50 steps.
+
+"""
+
+loss_fn = tf.keras.losses.MeanAbsoluteError()
+optim_edsr = tf.keras.optimizers.Adam(
+    learning_rate=PiecewiseConstantDecay(boundaries=[50], values=[1e-4, 5e-5])
+)
+
+
+@tf.function
+def train_step(ds_low, ds_high):
```

Comment:
Let's refactor the model as a `Model` subclass with a custom `train_step(0)` method. https://keras.io/guides/customizing_what_happens_in_fit/